File: backend/app/api/endpoints/applications.py
"""
API endpoints for job applications.
"""
from typing import List
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.api.deps import get_db, get_current_user
from app.models.user import User
from app.schemas.job_application import (
    JobApplication,
    JobApplicationCreate,
    JobApplicationUpdate,
    JobApplicationMove,
)
from app.services.job_service import JobApplicationService

router = APIRouter()


@router.get("/", response_model=List[JobApplication])
def list_applications(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    List all job applications for the current user.

    Returns:
        List of all job applications for current user
    """
    return JobApplicationService.get_all(db, current_user.id)


@router.post("/", response_model=JobApplication, status_code=status.HTTP_201_CREATED)
def create_application(
    application: JobApplicationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Create a new job application for the current user.

    Args:
        application: Job application data

    Returns:
        Created job application
    """
    return JobApplicationService.create(db, application, current_user.id)


@router.get("/{application_id}", response_model=JobApplication)
def get_application(
    application_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get a single job application by ID for the current user.

    Args:
        application_id: Application UUID

    Returns:
        Job application

    Raises:
        HTTPException: If application not found or doesn't belong to user
    """
    application = JobApplicationService.get_by_id(db, application_id, current_user.id)
    if not application:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Application {application_id} not found",
        )
    return application


@router.put("/{application_id}", response_model=JobApplication)
def update_application(
    application_id: UUID,
    application: JobApplicationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Update a job application for the current user.

    Args:
        application_id: Application UUID
        application: Updated data

    Returns:
        Updated job application

    Raises:
        HTTPException: If application not found or doesn't belong to user
    """
    updated = JobApplicationService.update(db, application_id, application, current_user.id)
    if not updated:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Application {application_id} not found",
        )
    return updated


@router.delete("/{application_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_application(
    application_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Delete a job application for the current user.

    Args:
        application_id: Application UUID

    Raises:
        HTTPException: If application not found or doesn't belong to user
    """
    deleted = JobApplicationService.delete(db, application_id, current_user.id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Application {application_id} not found",
        )


@router.patch("/{application_id}/move", response_model=JobApplication)
def move_application(
    application_id: UUID,
    move_data: JobApplicationMove,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Move application to a new status (for drag-drop) for the current user.

    Args:
        application_id: Application UUID
        move_data: New status and position

    Returns:
        Updated job application

    Raises:
        HTTPException: If application not found or doesn't belong to user
    """
    moved = JobApplicationService.move(db, application_id, move_data, current_user.id)
    if not moved:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Application {application_id} not found",
        )
    return moved

File: backend/src/routes/applications.ts
/**
 * Job applications API routes
 */
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { HTTPException } from 'hono/http-exception';
import { ZodError } from 'zod';
import { authMiddleware, type AuthContext } from '../middleware/auth.js';
import {
  getAllApplications,
  getApplicationById,
  createApplication,
  updateApplication,
  deleteApplication,
  moveApplication,
} from '../services/job-service.js';
import {
  jobApplicationCreateSchema,
  jobApplicationUpdateSchema,
  jobApplicationMoveSchema,
} from '../schemas/job-application.js';
import { logger } from '../lib/logger.js';

const applications = new Hono<AuthContext>();

// Apply auth middleware to all routes
applications.use('/*', authMiddleware);

/**
 * List all job applications for the current user
 * GET /applications
 */
applications.get('/', async (c) => {
  const user = c.get('user');

  try {
    logger.debug({
      userId: user.id
    }, 'Fetching all job applications');

    const apps = await getAllApplications(user.id);

    logger.debug({
      userId: user.id,
      count: apps.length
    }, 'Fetched job applications successfully');

    return c.json(apps);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id
    }, 'Failed to fetch job applications');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

/**
 * Create a new job application for the current user
 * POST /applications
 */
applications.post('/', zValidator('json', jobApplicationCreateSchema, (result, c) => {
  if (!result.success) {
    logger.error({
      error: result.error,
      issues: result.error.issues,
      path: '/applications',
      method: 'POST'
    }, 'Validation failed for create application');

    return c.json({
      error: 'Validation failed',
      details: result.error.issues.map(issue => ({
        path: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    }, 400);
  }
}), async (c) => {
  const user = c.get('user');
  const applicationData = c.req.valid('json');

  try {
    logger.info({
      userId: user.id,
      data: applicationData
    }, 'Creating job application');

    const app = await createApplication(applicationData, user.id);

    logger.info({
      userId: user.id,
      applicationId: app.id
    }, 'Job application created successfully');

    return c.json(app, 201);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id,
      data: applicationData
    }, 'Failed to create job application');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

/**
 * Get a single job application by ID for the current user
 * GET /applications/:id
 */
applications.get('/:id', async (c) => {
  const user = c.get('user');
  const applicationId = c.req.param('id');

  try {
    logger.debug({
      userId: user.id,
      applicationId
    }, 'Fetching job application by ID');

    const app = await getApplicationById(applicationId, user.id);
    if (!app) {
      logger.warn({
        userId: user.id,
        applicationId
      }, 'Application not found');
      throw new HTTPException(404, { message: `Application ${applicationId} not found` });
    }

    logger.debug({
      userId: user.id,
      applicationId
    }, 'Fetched job application successfully');

    return c.json(app);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id,
      applicationId
    }, 'Failed to fetch job application');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

/**
 * Update a job application for the current user
 * PUT /applications/:id
 */
applications.put('/:id', zValidator('json', jobApplicationUpdateSchema, (result, c) => {
  if (!result.success) {
    logger.error({
      error: result.error,
      issues: result.error.issues,
      path: '/applications/:id',
      method: 'PUT'
    }, 'Validation failed for update application');

    return c.json({
      error: 'Validation failed',
      details: result.error.issues.map(issue => ({
        path: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    }, 400);
  }
}), async (c) => {
  const user = c.get('user');
  const applicationId = c.req.param('id');
  const applicationData = c.req.valid('json');

  try {
    logger.info({
      userId: user.id,
      applicationId,
      data: applicationData
    }, 'Updating job application');

    const updated = await updateApplication(applicationId, applicationData, user.id);
    if (!updated) {
      logger.warn({
        userId: user.id,
        applicationId
      }, 'Application not found for update');
      throw new HTTPException(404, { message: `Application ${applicationId} not found` });
    }

    logger.info({
      userId: user.id,
      applicationId
    }, 'Job application updated successfully');

    return c.json(updated);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id,
      applicationId,
      data: applicationData
    }, 'Failed to update job application');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

/**
 * Delete a job application for the current user
 * DELETE /applications/:id
 */
applications.delete('/:id', async (c) => {
  const user = c.get('user');
  const applicationId = c.req.param('id');

  try {
    logger.info({
      userId: user.id,
      applicationId
    }, 'Deleting job application');

    const deleted = await deleteApplication(applicationId, user.id);
    if (!deleted) {
      logger.warn({
        userId: user.id,
        applicationId
      }, 'Application not found for deletion');
      throw new HTTPException(404, { message: `Application ${applicationId} not found` });
    }

    logger.info({
      userId: user.id,
      applicationId
    }, 'Job application deleted successfully');

    return c.body(null, 204);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id,
      applicationId
    }, 'Failed to delete job application');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

/**
 * Move application to a new status (for drag-drop) for the current user
 * PATCH /applications/:id/move
 */
applications.patch('/:id/move', zValidator('json', jobApplicationMoveSchema, (result, c) => {
  if (!result.success) {
    logger.error({
      error: result.error,
      issues: result.error.issues,
      path: '/applications/:id/move',
      method: 'PATCH'
    }, 'Validation failed for move application');

    return c.json({
      error: 'Validation failed',
      details: result.error.issues.map(issue => ({
        path: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    }, 400);
  }
}), async (c) => {
  const user = c.get('user');
  const applicationId = c.req.param('id');
  const moveData = c.req.valid('json');

  try {
    logger.info({
      userId: user.id,
      applicationId,
      moveData
    }, 'Moving job application');

    const moved = await moveApplication(applicationId, moveData, user.id);
    if (!moved) {
      logger.warn({
        userId: user.id,
        applicationId
      }, 'Application not found for move');
      throw new HTTPException(404, { message: `Application ${applicationId} not found` });
    }

    logger.info({
      userId: user.id,
      applicationId,
      newStatus: moveData.status
    }, 'Job application moved successfully');

    return c.json(moved);
  } catch (error) {
    logger.error({
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      userId: user.id,
      applicationId,
      moveData
    }, 'Failed to move job application');

    if (error instanceof HTTPException) {
      throw error;
    }
    throw new HTTPException(500, { message: 'Internal server error' });
  }
});

export default applications;
